test/hotspot/jtreg/sources/TestNoNULL.java

```diff
@@ -0,0 +1,146 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class TestNoNULL {
+    private static final List<Path> files = new ArrayList<>();
+    private static final Set<String> excludedSourceFiles = new HashSet<>();
+    private static final Set<String> excludedTestFiles = new HashSet<>();
+    private static final Set<String> excludedTestExtensions = Set.of(".c", ".java", ".jar");
+    private static final Pattern NULL_PATTERN = Pattern.compile("\\bNULL\\b");
+    private static Path dir = Paths.get(System.getProperty("test.src"));
+
+    public static void main(String[] args) throws IOException {
+        int maxIter = 20;
+        while (maxIter-- > 0 && dir != null && !Files.exists(dir.resolve("src"))) {
+            dir = dir.getParent();
+        }
+
+        if (dir == null) {
+            throw new RuntimeException("Could not locate the 'src' directory within 20 parent directories.");
+        }
+
+        Path srcPath = dir.resolve("src").resolve("hotspot");
+        Path testPath = dir.resolve("test").resolve("hotspot");
+
+        initializeExcludedPaths(dir);
+
+        processFiles(srcPath, excludedSourceFiles, false);
+        processFiles(testPath, excludedTestFiles, true);
+
+        boolean foundNull = false;
+        for (Path file : files) {
+            if (checkForNull(file)) {
+                foundNull = true;
+            }
+        }
+
+        if (foundNull) {
+            throw new RuntimeException("Found usage of 'NULL' in source files. See errors above.");
+        }
+    }
+
+    private static void initializeExcludedPaths(Path rootDir) {
+        List<String> sourceExclusions = List.of(
+                "src/hotspot/share/prims/jvmti.xml",
+                "src/hotspot/share/prims/jvmti.xsl",
+                "src/hotspot/share/utilities/globalDefinitions_visCPP.hpp",
+                "src/hotspot/share/utilities/globalDefinitions_gcc.hpp",
+                "src/hotspot/os/windows/os_windows.cpp" //todo: remove after JDK-8349417
+        );
+
+        List<String> testExclusions = List.of(
+                "test/hotspot/jtreg/vmTestbase/nsk/share/jvmti/README",
+                "test/hotspot/jtreg/vmTestbase/nsk/share/jni/README"
+        );
+
+        sourceExclusions.forEach(relativePath ->  excludedSourceFiles.add(rootDir.resolve(relativePath).normalize().toString()));
+        testExclusions.forEach(relativePath ->  excludedTestFiles.add(rootDir.resolve(relativePath).normalize().toString()));
+    }
+
+    private static void processFiles(Path directory, Set<String> excludedFiles, boolean excludeExtensions) throws IOException {
+        Files.walkFileTree(directory, new SimpleFileVisitor<>() {
+            @Override
+            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+                if (isIncluded(file, excludedFiles, excludeExtensions)) {
+                    files.add(file);
```

Comment:
Why collect files to be checked, rather than just checking them here?

Reply:
I agree with @kimbarrett; there is no point creating buffer collections that consume more resources with no real gains.